shims/spark35/src/main/scala/org/apache/spark/sql/execution/adaptive/GlutenCostEvaluator.scala

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.catalyst.SQLConfHelper
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.joins.ShuffledJoin
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * This [[CostEvaluator]] is to force use the new physical plan when AQE's new physical plan
+ * contains [[ShuffledJoin]] and cost is the same.
+ */
+case class GlutenCostEvaluator() extends CostEvaluator with SQLConfHelper {
+  private var isReOptimize = false
+  override def evaluateCost(plan: SparkPlan): Cost = {
+    val forceOptimizeSkewedJoin = conf.getConf(SQLConf.ADAPTIVE_FORCE_OPTIMIZE_SKEWED_JOIN)
+    val simpleCost =
+      SimpleCostEvaluator(forceOptimizeSkewedJoin).evaluateCost(plan).asInstanceOf[SimpleCost]
+    var cost = simpleCost.value * 2
+    if (isReOptimize && plan.exists(_.isInstanceOf[ShuffledJoin])) {
```

Comment:
We should only change the plan when there is an actual better build side.

Reply:
I have considered it before, but doing this requires a lot of cost. First, need to add extra rule to add tags. Secondly, when the plan contains multiple joins, it is not easy to judge which join is about to be executed. I just modified cost equal. If the costs are equal, it is forced to use new physics plan, which doesn't seem to hurt.

Reply:
> If the costs are equal, it is forced to use new physics plan

Would you like to confirm why the newer plan is always better than the older one? E.g., The newer one carries the latest up-to-date statistics so the hash-to-sort conversion in columnar rule could be more reliable?

If we can make sure the newer is better, then it sounds reasonable to use the newer one.

Reply:
Additionally, I assume `reoptimizePlan == plan` happens frequently speaking of AQE execution. If so we should measure on the planer performance too.

Reply:
The default condition for spark to use the new physics plan is `newCost < origCost || (newCost == origCost && currentPhysicalPlan != newPhysicalPlan)` , this pr just adds a situation, `newCost == origCost && currentPhysicalPlan == newPhysicalPlan`, so I think the performance of newPhysicalPlan has not changed.